parquet-column/src/main/java/parquet/schema/Types.java

```diff
@@ -555,6 +607,342 @@ protected GroupType build(String name) {
           "Cannot build an empty group");
       return new GroupType(repetition, name, originalType, fields, id);
     }
+
+    public MapBuilder<Q> map(
+        Type.Repetition repetition) {
+      return new MapBuilder<Q>(self()).repetition(repetition);
+    }
+
+    public MapBuilder<Q> requiredMap() {
+      return new MapBuilder<Q>(self())
+          .repetition(Type.Repetition.REQUIRED);
+    }
+
+    public MapBuilder<Q> optionalMap() {
+      return new MapBuilder<Q>(self())
+          .repetition(Type.Repetition.OPTIONAL);
+    }
+
+    public MapBuilder<Q> repeatedMap() {
+      return new MapBuilder<Q>(self())
+          .repetition(Type.Repetition.REPEATED);
+    }
+
+    public ListBuilder<BaseGroupBuilder<P, Q>> list(Type.Repetition repetition) {
+      return new ListBuilder<BaseGroupBuilder<P, Q>>(this).repetition(repetition);
+    }
+
+    public ListBuilder<BaseGroupBuilder<P, Q>> requiredList() {
+      return list(Type.Repetition.REQUIRED);
+    }
+
+    public ListBuilder<BaseGroupBuilder<P, Q>> optionalList() {
+      return list(Type.Repetition.OPTIONAL);
+    }
+  }
+
+  /**
+   * A builder for {@link GroupType} objects.
+   *
+   * @param <P> The type that this builder will return from
+   *          {@link #named(String)} when the type is built.
+   */
+  public static class GroupBuilder<P> extends BaseGroupBuilder<P, GroupBuilder<P>> {
+
+    private GroupBuilder(P parent) {
+      super(parent);
+    }
+
+    private GroupBuilder(Class<P> returnType) {
+      super(returnType);
+    }
+
+    @Override
+    protected GroupBuilder<P> self() {
+      return this;
+    }
+  }
+
+  public static class MapBuilder<P> extends Builder<MapBuilder<P>, P> {
+    private static final Type STRING_KEY = Types
+        .required(PrimitiveTypeName.BINARY).as(OriginalType.UTF8).named("key");
+
+    public static class MapKeyBuilder<P, Q> extends PrimitiveBuilder<P> {
+      private final P parent;
+      private final Q grandParent;
+
+      public MapKeyBuilder(P parent, Q grandParent, PrimitiveTypeName type) {
+        super(parent, type);
+        this.parent = parent;
+        this.grandParent = grandParent;
+        repetition(Type.Repetition.REQUIRED);
+      }
+
+      @Override
+      public MapKeyBuilder<P, Q> as(OriginalType orig) {
+        return (MapKeyBuilder<P, Q>) super.as(orig);
```

Comment:
Rather than casting, this could just return `this` . . . but I think it would be cleaner to have the `BasePrimitiveBuilder` that I suggest elsewhere. The reason is that we don't want to have to call super and return this in a bunch of methods. This misses, for example, the `length` method on `PrimitiveBuilder`.